lib/dockerregistry/inventory.go

```diff
@@ -1531,6 +1572,152 @@ func (sc *SyncContext) GarbageCollect(
+// TODO: Maybe split this into 2 parts, so that each part can be unit-tested
+// separately (deletion of manifest lists vs deletion of other media types).
+func (sc *SyncContext) ClearRepository(
+	regName RegistryName,
+	mfest Manifest,
+	mkProducer func(RegistryContext, ImageName, Digest) stream.Producer,
+	customProcessRequest *ProcessRequest) {
+
+	// deleteRequstsPopulator returns a PopulateRequests that
+	// varies by a predicate. Closure city!
+	var deleteRequstsPopulator func(func(cr.MediaType) bool) PopulateRequests = func(predicate func(cr.MediaType) bool) PopulateRequests {
+
+		var populateRequests PopulateRequests = func(
+			sc *SyncContext,
+			reqs chan<- stream.ExternalRequest,
+			wg *sync.WaitGroup) {
+
+			for _, registry := range mfest.Registries {
+				// Skip over any registry that does not match the regName we want to
+				// wipe.
+				if registry.Name != regName {
+					continue
+				}
+				for imageName, digestTags := range sc.Inv[registry.Name] {
+					for digest, _ := range digestTags {
+						mediaType, ok := sc.DigestMediaType[digest]
+						if !ok {
+							fmt.Println("could not detect MediaType of digest", digest)
+							continue
+						}
+						if !predicate(mediaType) {
+							fmt.Printf("skipping digest %s mediaType %s\n", digest, mediaType)
+							continue
+						}
+						var req stream.ExternalRequest
+						req.StreamProducer = mkProducer(
+							registry,
+							imageName,
+							digest)
+						req.RequestParams = PromotionRequest{
+							Delete,
+							"",
+							registry.Name,
+							registry.ServiceAccount,
+
+							// No source image name, because tag deletions
+							// should only delete the what's in the
+							// destination registry
+							ImageName(""),
+
+							imageName,
+							digest,
+							"",
+							"",
+						}
+						wg.Add(1)
+						reqs <- req
+					}
+				}
+			}
+		}
+		return populateRequests
+	}
+
+	var processRequest ProcessRequest
+	var processRequestReal ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			jsons, errors := getJSONSFromProcess(req)
+			if len(errors) > 0 {
+				reqRes.Errors = errors
+				requestResults <- reqRes
+				// Don't skip over this request, because stderr here is ignorable.
+				// continue
+			}
+			for _, json := range jsons {
+				sc.Info("DELETED image:", json)
+			}
+			reqRes.Errors = errors
+			requestResults <- reqRes
+			wg.Add(-1)
+		}
+	}
+
+	captured := make(CapturedRequests)
+
+	if sc.DryRun {
+		processRequestDryRun := MkRequestCapturer(&captured)
+		processRequest = processRequestDryRun
+	} else {
+		processRequest = processRequestReal
+	}
+
+	if customProcessRequest != nil {
+		processRequest = *customProcessRequest
+	}
+
+	var isEqualTo (func(cr.MediaType) func(cr.MediaType) bool) = func(want cr.MediaType) func(cr.MediaType) bool {
```

Comment:
isEquallTo is currently a function that takes a cr.MediaType that returns a function that takes a cr.MediaType that returns a bool. Please also split this out.

Same as below.

Reply:
While these anonymous functions are independent, their names shouldn't pollute the namespace of this module. This is because they are not used by any other function in this module.

Can we defer refactoring it to later?